File: algorithms/code/leetcode/lc005_longest_palindromic_substring/lc005_longest_palindromic_substring.py
#!/usr/bin/env python3
# https://leetcode.com/problems/longest-palindromic-substring/
# See also: Manacher algorithm

import unittest


class Solution:
    def longestPalindrome(self, s: str) -> str:
        longest = ''
        index = 0
        while index < len(s):
            # Find head & tail
            head = tail = index
            while tail + 1 < len(s):
                if s[head] != s[tail + 1]:
                    break
                tail += 1
            index += 1 + tail - head
            # Expand
            while head - 1 >= 0 and tail + 1 < len(s):
                if s[head - 1] != s[tail + 1]:
                    break
                head -= 1
                tail += 1
            # Check
            if len(longest) < 1 + tail - head:
                longest = s[head : tail + 1]
        return longest


class TestCode(unittest.TestCase):
    def test_1(self):
        self.assertEqual(Solution().longestPalindrome('babad'), 'bab')

    def test_2(self):
        self.assertEqual(Solution().longestPalindrome('cbbd'), 'bb')

    def test_3(self):
        self.assertEqual(Solution().longestPalindrome('a'), 'a')

    def test_4(self):
        self.assertEqual(Solution().longestPalindrome('ac'), 'a')

    def test_bb(self):
        self.assertEqual(Solution().longestPalindrome('bb'), 'bb')

    def test_longer(self):
        self.assertEqual(
            # pylint: disable=C0301
            Solution().longestPalindrome(
                'civilwartestingwhetherthatnaptionoranynartionsoconceivedandsodedicatedcanlongendureWeareqmetonagreatbattlefiemldoftzhatwarWehavecometodedicpateaportionofthatfieldasafinalrestingplaceforthosewhoheregavetheirlivesthatthatnationmightliveItisaltogetherfangandproperthatweshoulddothisButinalargersensewecannotdedicatewecannotconsecratewecannothallowthisgroundThebravelmenlivinganddeadwhostruggledherehaveconsecrateditfaraboveourpoorponwertoaddordetractTgheworldadswfilllittlenotlenorlongrememberwhatwesayherebutitcanneverforgetwhattheydidhereItisforusthelivingrathertobededicatedheretotheulnfinishedworkwhichtheywhofoughtherehavethusfarsonoblyadvancedItisratherforustobeherededicatedtothegreattdafskremainingbeforeusthatfromthesehonoreddeadwetakeincreaseddevotiontothatcauseforwhichtheygavethelastpfullmeasureofdevotionthatweherehighlyresolvethatthesedeadshallnothavediedinvainthatthisnationunsderGodshallhaveanewbirthoffreedomandthatgovernmentofthepeoplebythepeopleforthepeopleshallnotperishfromtheearth'  # noqa
            ),
            'ranynar',
        )

File: algorithms/code/leetcode/lc005_longest_palindromic_substring/lc005_longest_palindromic_substring.rs
// https://leetcode.com/problems/longest-palindromic-substring/
// See also: Manacher algorithm

pub struct Solution;

impl Solution {
    pub fn longest_palindrome(s: String) -> String {
        let mut longest = "".to_string();
        let mut index = 0;
        while index < s.len() {
            // Find head & tail
            let mut head = index;
            let mut tail = index;
            while tail + 1 < s.len() && s.chars().nth(head) == s.chars().nth(tail + 1) {
                tail += 1;
            }
            index += 1 + tail - head;
            // Expand
            while head > 0
                && tail + 1 < s.len()
                && s.chars().nth(head - 1) == s.chars().nth(tail + 1)
            {
                head -= 1;
                tail += 1;
            }
            // Check
            if longest.len() < tail - head + 1 {
                longest = s[head..=tail].to_string();
            }
        }
        longest
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_1() {
        assert_eq!(Solution::longest_palindrome("babad".to_string()), "bab");
    }

    #[test]
    fn test_2() {
        assert_eq!(Solution::longest_palindrome("cbbd".to_string()), "bb");
    }

    #[test]
    fn test_3() {
        assert_eq!(Solution::longest_palindrome("a".to_string()), "a");
    }

    #[test]
    fn test_4() {
        assert_eq!(Solution::longest_palindrome("ac".to_string()), "a");
    }

    #[test]
    fn test_bb() {
        assert_eq!(Solution::longest_palindrome("bb".to_string()), "bb");
    }

    #[test]
    fn test_longer() {
        assert_eq!(
            Solution::longest_palindrome(
                "civilwartestingwhetherthatnaptionoranynartionsoconceivedandsodedicatedcanlongendureWeareqmetonagreatbattlefiemldoftzhatwarWehavecometodedicpateaportionofthatfieldasafinalrestingplaceforthosewhoheregavetheirlivesthatthatnationmightliveItisaltogetherfangandproperthatweshoulddothisButinalargersensewecannotdedicatewecannotconsecratewecannothallowthisgroundThebravelmenlivinganddeadwhostruggledherehaveconsecrateditfaraboveourpoorponwertoaddordetractTgheworldadswfilllittlenotlenorlongrememberwhatwesayherebutitcanneverforgetwhattheydidhereItisforusthelivingrathertobededicatedheretotheulnfinishedworkwhichtheywhofoughtherehavethusfarsonoblyadvancedItisratherforustobeherededicatedtothegreattdafskremainingbeforeusthatfromthesehonoreddeadwetakeincreaseddevotiontothatcauseforwhichtheygavethelastpfullmeasureofdevotionthatweherehighlyresolvethatthesedeadshallnothavediedinvainthatthisnationunsderGodshallhaveanewbirthoffreedomandthatgovernmentofthepeoplebythepeopleforthepeopleshallnotperishfromtheearth".to_string()
            ),
            "ranynar",
        );
    }
}
